src/Roslyn.Diagnostics.Analyzers/Core/SymbolIsBannedAnalyzer.cs

```diff
@@ -0,0 +1,314 @@
+                ? SymbolDisplayFormat.CSharpShortErrorMessageFormat
+                : SymbolDisplayFormat.VisualBasicShortErrorMessageFormat;
+
+            var bannedAttributes = bannedSymbols
+                .Where(s => s.symbol is ITypeSymbol n && n.IsAttribute())
+                .ToImmutableDictionary(s => s.symbol, s => s.message);
+
+            List<Diagnostic> diagnostics = null;
+
+            if (bannedAttributes.Count > 0)
+            {
+                compilationContext.RegisterCompilationEndAction(
+                    context =>
+                    {
+                        VerifyAttributes(compilationContext.Compilation.Assembly.GetAttributes());
+                        VerifyAttributes(compilationContext.Compilation.SourceModule.GetAttributes());
+                    });
+
+                compilationContext.RegisterSymbolAction(
+                    context => VerifyAttributes(context.Symbol.GetAttributes()),
+                    SymbolKind.NamedType,
+                    SymbolKind.Method,
+                    SymbolKind.Field,
+                    SymbolKind.Property,
+                    SymbolKind.Event);
+            }
+
+            compilationContext.RegisterOperationAction(
+                context =>
+                {
+                    switch (context.Operation)
+                    {
+                        case IObjectCreationOperation objectCreation:
+                            VerifySymbol(objectCreation.Constructor, context.Operation.Syntax);
+                            VerifyType(objectCreation.Type.OriginalDefinition, context.Operation.Syntax);
+                            break;
+
+                        case IInvocationOperation invocation:
+                            VerifySymbol(invocation.TargetMethod, context.Operation.Syntax);
+                            VerifyType(invocation.TargetMethod.ContainingType.OriginalDefinition, context.Operation.Syntax);
+                            break;
+
+                        case IMemberReferenceOperation memberReference:
+                            VerifySymbol(memberReference.Member, context.Operation.Syntax);
+                            VerifyType(memberReference.Member.ContainingType.OriginalDefinition, context.Operation.Syntax);
+                            break;
+                    }
+                },
+                OperationKind.ObjectCreation,
+                OperationKind.Invocation,
+                OperationKind.EventReference,
+                OperationKind.FieldReference,
+                OperationKind.MethodReference,
+                OperationKind.PropertyReference);
+
+            if (compilationContext.Compilation.Language == LanguageNames.CSharp)
+            {
+                compilationContext.RegisterSyntaxNodeAction(
+                    context => VerifyDocumentationSyntax(((Microsoft.CodeAnalysis.CSharp.Syntax.XmlCrefAttributeSyntax)context.Node).Cref, context),
```

Comment:
This violates the layering contracts - the core layer should not reference any language specific syntax nodes or kinds. Additionally, this will cause VB analyzers and code analysis assemblies to be loaded from a C# project and vice versa, which has negative performance implications.

Recommended approach is to make this an abstract analyzer type, with C# and VB specific sub-types. For example, you can have an abstract property `TSyntaxKind XmlCrefSyntaxKind {get;}` in the core analyzer and your registration can be simplified to a single language agnostic registration:
`compilationContext.RegisterSyntaxNodeAction(VerifyDocumentationSyntax, XmlCrefSyntaxKind);`, which shouldn't need any casts.

Reply:
Thanks, TIL. Should be fixed in 5de90322dd79a2e608224664128006e5653e6ad2.